File: c/xover.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))

const unsigned ITERATIONS = 100000;
const unsigned LENGTH	 =  32768;

int main()
{
	for (unsigned length = 16; length <= LENGTH; length <<= 1)
	{
		struct timespec start, stop;
		char bits1[length], bits2[length];

		for (unsigned i = 0; i < length; i++ )
		{
			bits1[i] = rand() & 1;
			bits2[i] = rand() & 1;
		}

		clock_gettime(CLOCK_REALTIME, &start);

		// two point crossover
		for (unsigned i = 0; i < ITERATIONS; ++i)
		{
			unsigned t1 = rand() & (length - 1), 
			         t2 = rand() & (length - 1);
			unsigned b = MIN(t1, t2), 
					 e = MAX(t1, t2);
			while (b != e)
			{
				char tmp = bits1[b];
				bits1[b] = bits2[b];
				bits2[b] = tmp;
				++b;
			}
		}
		
		clock_gettime(CLOCK_REALTIME, &stop);

		printf("C-char[], %u, %f\n", length, (stop.tv_sec - start.tv_sec) + (stop.tv_nsec - start.tv_nsec) / 1e9);
	}

	return 0;
}

File: csharp/xover.cs
using System;
using System.Diagnostics;
using System.Linq;

namespace xover
{
	class MainClass
	{
		const uint ITERATIONS = 100000;
		const ushort LENGTH     =  ushort.MaxValue;
		static Random rnd = new Random();

		public static void Main (string[] args)
		{

			for (ushort length = 16; length != 0 ; length <<= 1) {


				bool[] bits1 = new bool[length];
				bool[] bits2 = new bool[length];

				for (uint i = 0; i < length; i++) {
					bits1[i] = Convert.ToBoolean (rnd.Next (0, 2));
					bits2[i] = Convert.ToBoolean (rnd.Next (0, 2));

				}

				Stopwatch stopwatch = Stopwatch.StartNew (); 

				for (int i = 0; i < ITERATIONS; i++) {
					
				


					//  TWO POINT
					int point1 =  rnd.Next (length);
					int point2 =  rnd.Next (length);
					int min = Math.Min (point1, point2);
					int max = Math.Max (point1, point2);


					for (int j = min; j <= max; j++){
						
							bool tmp = bits1 [j];
							bits1 [j] = bits2 [j];
							bits2 [j] = tmp;

						}

				}
				    

				stopwatch.Stop ();

				long ticks = stopwatch.ElapsedTicks;
				double s = (double)ticks / Stopwatch.Frequency;

				Console.WriteLine ("{0},{1}",length,s);
			}
		}
	}
}